### ISSUE TITLE
[Feature Request] PR Velocity-related dashboards for Kubernetes

### ISSUE THREAD

**Issue by lasomethingsomething** [cncf/devstats.archive#248]
Hi there: Thank you for the previous set of changes to the Kubernetes dashboards. I am wondering if these could be added—most of them are meant to measure pull request velocity per SIG in real-time, instead of the manual point-in-time data collection I conducted a few weeks ago (and shared the results of via spreadsheet):

- Number of open PRs by SIG (comparative chart)
- Number of PRs open longer than 30 days, 60 days, 90 days by SIG (comparative chart)
- Number of open `needs-priority` issues and PRs, by SIG
- Number of open `priority/critical-urgent` by SIG (comparative chart)
- Number of open PRs with `needs-rebase` label by SIG (comparative chart)
- Number of currently *blocked* PRs by SIG (comparative chart); there is [this dashboard](https://k8s.devstats.cncf.io/d/4/blocked-prs-repository-groups?orgId=1&var-period=m&var-repogroup=%22All%22), but requires a SIG-by-SIG check to get specifics
- Number of "no activity on an issue or PR" items in 2 and +2 weeks, per SIG

Also: `PM` is no longer a SIG and can be removed from the overall list of SIGs. 

Please let me know if you have any questions. Thank you!

**Comment by lukaszgryglicki** [comment]
I'll review this one by one and will see what can I do. Definitely, all this data can be gathered just once, the question is does the community wants to have such dashboards added to official ones?
cc @jberkus @dankohn ?

**Comment by lasomethingsomething** [comment]
Thank you, @lukaszgryglicki. Some of us pinged on this thread discussed this issue during the Contribex biweekly update just now. Some notes:

- These dashboards, and the metrics they collect, are great and useful 
- However, feedback from SIG Chairs and Leads about them has been light, despite numerous requests for feedback in the past
- Based on the lack of feedback, we perceive that the SIG Chairs and Leads are not leveraging devstats as part of their SIGs' regular workflow and operations
- It would be great to increase devstats adoption among the Chairs/Leads group

On the last point: One use case for the devstats is around community management and understanding the contribution landscape. But others might be:
- I'm part of Contribex, the Steering Committee, or some other group that's trying to assess workflow for the K8s project. I want a SIG-by-SIG view of some key (and potential) diagnostic metrics so that I can lend support to help SIGs in need reduce or balance workload.
- I participate in a SIG (either in a leadership position or as a regular contributor) and I want to understand our workflow better, in relation to other SIGs.

For these other two use cases, the data is all there. But unless I'm missing something, some of the requested metrics like "# of blocked PRs" are currently collectable on devstats only by clicking repository group by repository group; downloading the data for a given time frame (7 days, month, etc.); and then manually placing the SIG data in a spreadsheet for comparison. This data, captured in a single chart in real time, would efficiently support the two use cases above.

The other suggested metrics are meant to measure workflow lag time from different angles. Again, I think much of the data is already there, but either needs some additional teasing out or reformatting.

I hope this is helpful!

**Comment by lukaszgryglicki** [comment]
So I can create such dashboards, but currently, I'm a bit busy and must postpone this a bit.

**Comment by lukaszgryglicki** [comment]
Working on this.

**Comment by lukaszgryglicki** [comment]
So:
- `Also: PM is no longer a SIG and can be removed from the overall list of SIGs.` - OK but not that DevStats is supposed to show *history* not only the current state, so it if *was* a SIG some time ago I would leave it.
- `Number of currently blocked PRs by SIG (comparative chart); there is this dashboard, but requires a SIG-by-SIG check to get specifics` - so I guess I need to wait for some final feedback in this case, right?
- `Number of "no activity on an issue or PR" items in 2 and +2 weeks, per SIG` - just to confirm - you want a number of PRs and issues inactive in 2 weeks (how many issues and PRs were inactive during the last 2 weeks, excluding bot activity) and what do that +2 means- inactive in what period?

**Comment by lasomethingsomething** [comment]
- On `PM`: Sounds good. But yes, PM is no longer active as of April 2020.

- `Number of currently blocked...`: I'm not sure what feedback is needed. Essentially the goal is to show all SIGs' blocked PRs in one chart. IIRC these are the labels currently captured in the blocked PR data:
      - All open PRs at that time
      - Needs OK to test
      - Release note label needed
      - No LGTM label
      - No approved label
      - Do not merge label(s)
This is also useful info. Would be great to have a general dashboard showing all blocked PRs SIG-by-SIG in one chart/view, then a way (possibly a drop-down) to drill down further into the data using the above label-queries. 

- `Number of "no activity...`: Yes. Would be great to be able to separate PRs from Issues as well. After getting some feedback from SIG Chairs and TLs, a more useful set of dates would look like: not updated by a human in +2 weeks | +30 days | +90 days. 

A note on time: I was having trouble fully understanding how the date ranges worked on the dashboards. When I tried to use the quarterly ranges, I'd get unlikely results for the current quarter. I read the write-ups for the dashboards but still didn't quite get it. So I did a workaround and did month-by-month data collection (downloading the data) instead.

**Comment by lukaszgryglicki** [comment]
OK will see some first preview soon - now clear on +2 weeks, +30 days and +90 days, thanks.

**Comment by lukaszgryglicki** [comment]
Generally, date ranges are "aggregation periods" so for example quarter means data is aggregated over that quarter when you see PRs with "period" = "quarter" you will see the number of PRs opened in each quarter. There are also MA periods. MA = moving average. In that case (for example 7 Days MA) you have a "7 days long" window moving day by day and averaging results from that window (so usually dividing results by 7 for every window). That helps to eliminate weekends impact on charts shape, similar 24 Hours MA etc.
I should have some first previews tomorrow.

**Comment by lukaszgryglicki** [comment]
**OK** so showing first few test previews (only running on the test server so URL address contains `teststats` instead of `devstats`):
- Number of open PRs by SIG (comparative chart) - [click](https://k8s.teststats.cncf.io/d/69/prs-opened-by-sig?orgId=1).
- Number of PRs open longer than 30 days, 60 days, 90 days by SIG (comparative chart) - [click](https://k8s.teststats.cncf.io/d/70/awaiting-prs-by-sig?orgId=1) (added 10 days and 1 year in addition to 30, 60, 90 days - 10 days just show the recent/opened a bit longer than usual PR info, year shows permanently stalled PRs that should just be closed IMHO).

For the next few dashboards requesting PR labels per SIG I've created one, that allows choosing labels already used in [this dashboard](https://k8s.teststats.cncf.io/d/47/prs-labels-repository-groups?orgId=1) (also added there) in addition to `needs-priority`, `needs-rebase` and `priority/critical-urgent`.

- Number of open needs-priority issues and PRs, by SIG - [click](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=needs-priority).
- Number of open priority/critical-urgent by SIG (comparative chart) - [click](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=priority%2Fcritical-urgent).
- Number of open PRs with needs-rebase label by SIG (comparative chart) - [click](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=needs-rebase)

Remaining ones:
- Number of currently blocked PRs by SIG (comparative chart); there is [this dashboard](https://k8s.devstats.cncf.io/d/4/blocked-prs-repository-groups?orgId=1&var-period=m&var-repogroup=%22All%22), but requires a SIG-by-SIG check to get specifics - **just to clarify - you want a similar dashboard but with SIGs drop-down, right?**

- Number of "no activity on an issue or PR" items in 2 and +2 weeks, per SIG - not yet started.

**Comment by lasomethingsomething** [comment]
Thank you so much. A few notes:

- I did a test of the `priority/critical-urgent` label [here](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=priority%2Fcritical-urgent) and the result is 18, but when I do a [GH query](https://github.com/kubernetes/kubernetes/issues?q=is%3Aissue+is%3Aopen+label%3Apriority%2Fcritical-urgent) on k/k, for `is:issue`, it's 23 (all issues there have `sig-` labels).  For `is:pr`[(here](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+is%3Aopen+label%3Apriority%2Fcritical-urgent) it's 9 results. Wondering what might be happening?

- [Here](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=priority%2Fcritical-urgent), some of the charts are titled "percent PRs..." but the numbers in the right-hand table are 546, 655, etc. Wondering if that's an error or if I'm reading it incorrectly?

- For dashboards [like this one](https://k8s.teststats.cncf.io/d/69/prs-opened-by-sig?orgId=1&var-period=d7&var-sigs=All), on the right-hand side table, do the columns of data `avg` and `current` mean, `avg` = the average number of PRs in the selected time frame, and `current` = the current number of PRs in the selected time frame?

- +1 for organising [this `PR labels` dashboard](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=needs-rebase) like you did.

- Number of currently blocked PRs by SIG (comparative chart); there is this dashboard, but requires a SIG-by-SIG check to get specifics **- just to clarify - you want a similar dashboard but with SIGs drop-down, right?** Ah, I meant a dashboard like [the top one here](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=needs-rebase), where the SIGs are all placed together. Does that help?

Greatly appreciate how quickly you've helped here.

**Comment by lukaszgryglicki** [comment]
Will do next iteration tomorrow and then next and so on...
By "percent" it mean that stacked values sum to 100%, but yes, I'll change that - this is just a chart "shape" to sum all SIGs.

**Comment by lukaszgryglicki** [comment]
Regarding this:
```
I did a test of the priority/critical-urgent label here and the result is 18, but when I do a GH query on k/k, for is:issue, it's 23 (all issues there have sig- labels). For is:pr(here it's 9 results. Wondering what might be happening?
```
New dashboards only take PRs into account (as requested). So your query for `is:issue` has nothing to do here. Now, your query is for k/k repo, while my dashboards show all repos from "Kubernetes" project - which means at leats 200 repos not just k/k.
I'll do a very detailed analysis to see if my data is OK tomorrow (before proceeding to next dashboards) and will show you the exact repos why my dashboard shopws "17" - but now my guess is that 9 are from k/k and remaining 8 from other repos.
cc @LappleApple

**Comment by lasomethingsomething** [comment]
Thanks for checking in on above. Wasn't sure if data was coming from all repos or just k/k (all repos is +1), and good to know that it's filtering out issues as well.

**Comment by lukaszgryglicki** [comment]
```
For dashboards like this one, on the right-hand side table, do the columns of data avg and current mean, avg = the average number of PRs in the selected time frame, and current = the current number of PRs in the selected time frame?
```
Those columns are default Grafana columns - `avg` means avg of data from the chart for given column, `max` is maximum from datapoints, min is the minimum and `current` is the last data available (last timestamp - most recent).  

So for the number of currently blocked PRs [this](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=needs-rebase) (that dashboard was added yesterday) is what you need -it allows selecting a single label and then "All SIGs combined" and see a summary data for all SIGs, data for each separate SIG, or select "All Labels combined" and then see number of PRs blocked due to all those labels per SIG, basically I think all things you asked for, right?
Example: [SIG testing](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=testing&var-label_name=All%20labels%20combined) blocked PRs status, PRs blocked on [needs-priority](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=needs-priority) in all SIGs, [summary](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=All%20SIGs%20combined&var-label_name=All%20labels%20combined) for all labels and SIGs, [SIG node work-in-progress](https://k8s.teststats.cncf.io/d/71/prs-labels-by-sig?orgId=1&var-sig_name=node&var-label_name=do-not-merge%2Fwork-in-progress) status.

I'm checking this (to make sure the data is consistent, I'll see which repos report those PRs and the check with GitHub UI, sum and see if data matches):
```
I did a test of the priority/critical-urgent label here and the result is 18, but when I do a GH query on k/k, for is:issue, it's 23 (all issues there have sig- labels). For is:pr(here it's 9 results. Wondering what might be happening?
```

**Comment by lasomethingsomething** [comment]
Hiya: I did a brief presentation on DevStats last night for the SIG Chairs and Leads, and based on the feedback it seems that we might increase our chances of getting the group to see these tools as useful instead of nice-to-have by `prioritizing k/k for the breakdown`. 

I still think the `repo group` tallies are valuable, because they show overall SIG workload. But if k/k-specific stats will inspire Chairs, TLs, and others (release team members, for example) to use DevStats more regularly, then it's worth trying out the change. 

A few other suggestions/outcomes from the presentation and follow-up discussion:
- SIG subprojects all have their own set of cadence. Maybe a dashboard display by sig for their individual repos + k/k? This way, SIGs could see at a glance if a subproject is taking forever to respond, which would trigger a datapoint for Chairs/TLs to get more owners on that subproject.
- If the above point requires a ton of work, then perhaps only the SIGs with the heaviest workloads would get them? These would be api-machinery, node, testing, and apps.
- One SIG lead asked if it would be possible to measure the success of independent contributors in getting quick turnaround/responses, to see if they are responded-to as quickly as contributors from, say, noted/heavily invested companies. So maybe this could mean adding an "independent contributor" field to existing community health/assessment dashes.
- One SIG Chair suggested adapting the PR age data to better reveal if SIG workloads flatline/stagnate: ie, if a SIG never drops below a certain # of PRs blocked/still open at any given time, it's a signal for them to examine their processes.

**Comment by lukaszgryglicki** [comment]
Finished next round:

1) I've checked example with GitHub query you mentioned vs. the number of open PRs from the dashboard. Found one bug and regenerated data. My comment about counting all open PRs across Kubernetes (not only `k/k` repo) still applies.

2) Requested dashboard showing per-SIG number of issues inactive for 2 weeks +, 30 days +, 90 days + - [click](https://k8s.teststats.cncf.io/d/73/inactive-issues-by-sig?orgId=1&var-sigs=All). I'm counting the time since issue creation to first non-author and non-bot activity on the issue.

3) Requested dashboard showing per-SIG number of PRs inactive for 2 weeks +, 30 days +, 90 days + - [click](https://k8s.teststats.cncf.io/d/72/inactive-prs-by-sig?orgId=1). I'm counting the time since PR creation to first non-author and non-bot activity on the PR.

Regarding your newest [feedback](https://github.com/cncf/devstats/issues/248#issuecomment-627891984), I'm open to suggestions:
- what exact dashboard(s) should I add (like what columns, charts, drop-downs do you want to see there) - if any.
- what should I modify on the existing dashboard(s) - if any?

**Comment by jberkus** [comment]
@LappleApple are any existing dashboards specifically cited as *not* useful?   Looking to prune here.

**Comment by lasomethingsomething** [comment]
Not cited to me personally. I could give my own recs but I think we'd prefer to have the Chairs/TLs' general assessment, no?

**Comment by lukaszgryglicki** [comment]
Before I start to address the newest [feedback](https://github.com/cncf/devstats/issues/248#issuecomment-627891984) (not yet fully specified, maybe after today's K8s call?) should I add 5 new dashboards implemented here to `prod`?

**Comment by lasomethingsomething** [comment]
@lukaszgryglicki I'm not sure who normally answers this for you, do you need me to ping someone?

**Comment by lukaszgryglicki** [comment]
Hard to say, actually they're new so they should not harm, I'll add them to prod tomorrow, unless somebody objects. I can delete them anytime if needed.

**Comment by lasomethingsomething** [comment]
Thanks. I don't foresee any reason for us to delete them anytime soon. :)

**Comment by lukaszgryglicki** [comment]
All new dashboards ported to [prod](https://k8s.devstats.cncf.io). Will need to switch to some other task for a while, sorry.

**Comment by lukaszgryglicki** [comment]
Closing this, please reopen if needed.
